conf/supervisord-dev.conf

```diff
@@ -0,0 +1,35 @@
+[supervisord]
+nodaemon = true
+silent = true
+
+[program:web]
+command=gunicorn -c conf/gunicorn/dev.conf.py --paste conf/development.ini
+stdout_events_enabled=true
+stderr_events_enabled=true
+stopsignal = KILL
+stopasgroup = true
+
+[program:nginx]
+command=docker-compose up --no-color --exit-code-from nginx-proxy nginx-proxy
```

Comment:
This is slightly more complicated than just `docker-compose up`:

1. We only want to run the NGINX container (named `nginx-proxy`), and not any other service containers that might be defined in `docker-compose.yml` in future. Hence `docker-compose up ... nginx-proxy' (in fact `nginx-proxy` is the only container in `docker-compose.yml at the moment). To start just one container and also print its output it seems that you need to do `docker-compose up -

2. Added `--no-color` to let Supervisor (specifically: our `bin/logger` script) colorize the output not docker-compose

3 . Added `--exit-code-from` to tell docker-compose to proxy nginx's exit code

There are a couple of problems with this that're caused by using docker-compose to run nginx instead of running docker directly. These are pre-existing problems, for another PR:

1. Nginx is still left running after you <kbd><kbd>Ctrl</kbd> + <kbd>c</kbd></kbd> the `make dev` process. This isn't the right behaviour for `make dev` which should kill all the processes when stopped (and does so for every other process across our apps, except this one)

2. When you run `make dev` it starts by printing out a potentially long number of log lines from nginx from previous runs of `nginx`

Reply:
You probably want to take the same approach as `superdev` which has solved the closing on exit issue: 

* https://github.com/hypothesis/superdev/blob/master/conf/via3/services.conf

I would strongly suggest checking `superdev` as a first resort when attempting to work out how to run these.

Reply:
There is a `docker-compose run` command which can run a docker-compose service in a new container. With the following change I got `make dev` to start an nginx container which was then killed when `make dev` was interrupted. Making this second part work required changing the signal.

```diff
diff --git a/conf/supervisord-dev.conf b/conf/supervisord-dev.conf
index 12622bb..bb5cd8e 100644
--- a/conf/supervisord-dev.conf
+++ b/conf/supervisord-dev.conf
@@ -10,10 +10,10 @@ stopsignal = KILL
 stopasgroup = true
 
 [program:nginx]
-command=docker-compose up --no-color --exit-code-from nginx-proxy nginx-proxy
+command=docker-compose run --rm --service-ports nginx-proxy
 stdout_events_enabled=true
 stderr_events_enabled=true
-stopsignal = KILL
+stopsignal = INT
 stopasgroup = true
 
 [eventlistener:logger]
```

Reply:
Regarding the `stopsignal` change above, I wonder whether `SIGKILL` may be too aggressive for other processes as well by preventing them from doing cleanup that they otherwise would do.

Reply:
I had tried `--abort-on-container-exit` (as Superdev does) but it doesn't do what we want:

```terminal
$ git diff
diff --git a/conf/supervisord-dev.conf b/conf/supervisord-dev.conf
index 12622bb..9eec592 100644
--- a/conf/supervisord-dev.conf
+++ b/conf/supervisord-dev.conf
@@ -10,7 +10,7 @@ stopsignal = KILL
 stopasgroup = true
 
 [program:nginx]
-command=docker-compose up --no-color --exit-code-from nginx-proxy nginx-proxy
+command=docker-compose up --no-color --exit-code-from nginx-proxy --abort-on-container-exit nginx-proxy
 stdout_events_enabled=true
 stderr_events_enabled=true
 stopsignal = KILL
$ # Nothing running:
$ tox -qqe dev --run-command 'docker-compose ps'
Name   Command   State   Ports
------------------------------
~/P/via3 supervisor> make dev
web (stderr)         | [2020-08-10 13:31:48 +0100] [3498512] [INFO] Starting gunicorn 20.0.4
web (stderr)         | [2020-08-10 13:31:48 +0100] [3498512] [INFO] Listening at: http://0.0.0.0:9082 (3498512)
web (stderr)         | [2020-08-10 13:31:48 +0100] [3498512] [INFO] Using worker: sync
web (stderr)         | [2020-08-10 13:31:48 +0100] [3498538] [INFO] Booting worker with pid: 3498538
web (stderr)         | [2020-08-10 13:31:48 +0100] [3498549] [INFO] Booting worker with pid: 3498549
web (stderr)         | [2020-08-10 13:31:48 +0100] [3498567] [INFO] Booting worker with pid: 3498567
nginx (stderr)       | Creating network "via3_default" with the default driver
web (stderr)         | [2020-08-10 13:31:48 +0100] [3498600] [INFO] Booting worker with pid: 3498600
web                  | Cache buster salt: f730d93b5919787dd79832a0244a648a
web                  | Cache buster salt: f730d93b5919787dd79832a0244a648a
nginx (stderr)       | Creating via3_nginx-proxy_1 ... 
web                  | Cache buster salt: f730d93b5919787dd79832a0244a648a

nginx (stderr)       | Creating via3_nginx-proxy_1 ... done
nginx (stderr)       |

nginx                | Attaching to via3_nginx-proxy_1
^C
ERROR: got KeyboardInterrupt signal
130! ~/P/via3 supervisor> # make dev has not killed nginx:
130! ~/P/via3 supervisor> tox -qqe dev --run-command 'docker-compose ps'
       Name                     Command               State                Ports
----------------------------------------------------------------------------------------------
via3_nginx-proxy_1   /bin/sh -c /etc/nginx/dev_ ...   Up      80/tcp, 127.0.0.1:9083->9083/tcp
$ 
```

I think `--abort-on-container-exit` tells Docker Compose to kill the _other containers_ (in Via 3's case, there aren't any other containers) when one container stops ("Stops all containers if any container was stopped"). It doesn't tell Docker Compose to kill the container(s) when the `docker-compose up` command itself is <kbd><kbd>Ctrl</kbd> + <kbd>c</kbd></kbd>'d. It has to do with multi-container Docker Compose situations, doesn't do anything in a single-container situatin like Via 3's

Reply:
@robertknight Yes! The `docker-compose run --rm ...` command seems to do what we want as long as it's combined with the `stopsignal = INT` change. It also solves two other problems with docker-compose:

1. It removes "double-enveloping of nginx's log messages because docker-compose prefixes them with `ginx-proxy_1  |` and then supervisor prefixes them again with `nginx                |` so they all come out like `nginx                | nginx-proxy_1  | ...`. With `docker-compose run` Docker Compose no longer prefixes things so they come out as we want: `nginx                | ...`

2. It also gets rid of the problem that `make dev` would start by printing out lots of Nginx logs messages from the _previous_ run of `make dev` so you get an irrelevant screenful of junk. Generally having a lot of stuff get printed out when `make dev` starts if something to avoid because it's annoying and it hides any errors or warnings in all the noise

So this is a great improvement, thanks

Reply:
Done